### PR TITLE
Android keystore experiment

### PR DESCRIPTION
**Pull request by Mugurell**
PR for starting the Android Keystore AC Experiment and logging the telemetry about the results of the operations.

Added jhugman as a reviewer for the first commit - starting the experiment based on a Nimbus experiment (not yet live)
Added Grisha as a reviewer for the mapping of the AC Facts to Fenix telemetry Events.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

File: app/metrics.yaml [hunk 1]
```diff
@@ -4538,3 +4538,57 @@ engine_tab:
...
+      - To be added
+    data_sensitivity:
+      - technical
+    notification_emails:
+      - fenix-core@mozilla.com
+    expires: "2021-09-01"
+  reset:
+    type: event
+    description: |
+      The secure prefs file for the experiment was cleared.
```
Comment (grigoryk) [inline app/metrics.yaml:4585]: I'd reword this: "An experiment failed, and was reset to run again in the future from a blank state."

File: app/metrics.yaml [hunk 2]
```diff
@@ -4538,3 +4538,57 @@ engine_tab:
       - fenix-core@mozilla.com
       - skaspari@mozilla.com
     expires: "2021-12-31"
+
+android_keystore_experiment:
+  experiment:
+    type: event
+    description: |
+      General data about the functionality/experiment.
```
Comment (grigoryk) [inline app/metrics.yaml:4546]: This will only be recorded in case of unexpected failures, so reword as such: "Records an instance of an unexpected failure during the experiment"

File: app/src/main/java/org/mozilla/fenix/components/metrics/Event.kt [hunk 3]
```diff
@@ -221,6 +222,46 @@ sealed class Event {
             get() = hashMapOf(TopSites.swipeCarouselKeys.page to page.toString())
     }
 
+    data class SecurePrefsExperiment(
```
Comment (grigoryk) [inline app/src/main/java/org/mozilla/fenix/components/metrics/Event.kt:225]: You'll need to change this classes to match the new events (see comments in metrics.yaml).

Make sure to set the extras explicitly, they'll be generated by glean from the metrics.yaml definitions (see code sample in https://mozilla.github.io/glean/book/user/metrics/event.html)

File: app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt [hunk 4]
```diff
@@ -750,6 +751,19 @@ private val Event.wrapper: EventWrapper<*>?
             { BannerOpenInApp.goToSettings.record(it) }
         )
 
+        is Event.SecurePrefsExperiment -> EventWrapper<NoExtraKeys>(
```
Comment (grigoryk) [inline app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt:754]: will need to change to match new events.

File: app/src/main/java/org/mozilla/fenix/components/metrics/MetricController.kt [hunk 5]
```diff
@@ -267,6 +268,29 @@ internal class ReleaseMetricController(
             }
             null
         }
+
+        Component.LIB_DATAPROTECT to SecurePrefsReliabilityExperiment.Companion.Actions -> {
+            when (item) {
+                // Map each type of experiment with a result.
```
Comment (grigoryk) [inline app/src/main/java/org/mozilla/fenix/components/metrics/MetricController.kt:274]: You don't need to do anything with enums. Just record `value` directly into an extra, and make sure to tease out exception classes from the Fact's metadata and pass these into corresponding extras, as well.

**Review by grigoryk — CHANGES_REQUESTED** [review]
Asked for a few changes around how we structure metrics, and at which point we run the experiment itself.

Also, I'm aware that this is introducing a level of "hidden" coupling between the a-c experiment code and the metrics being recorded here (e.g. we have to rely on alignment of structures, if that changes in a-c we won't catch the problems at the compile time), but: I think that was basically inevitable, we won't be changing the a-c code while this exists, and also and this code has a pretty distinct expiration, too (it doesn't need to exist after we've gathered the data).

File: app/src/main/java/org/mozilla/fenix/components/metrics/SecurePrefsTelemetry.kt [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    private val experiments: NimbusApi
+) {
+    fun startTests() {
+        // The Android Keystore is used to secure the shared prefs only on API 23+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            // These tests should run only if the experiment is live
+            experiments.withExperiment(Experiments.ANDROID_KEYSTORE) { experimentBranch ->
+                // .. and this device is not in the control group.
+                if (experimentBranch == ExperimentBranch.TREATMENT) {
+                    GlobalScope.launch {
```
[the remark this answers is not in the document]
Comment (grigoryk) [reply app/src/main/java/org/mozilla/fenix/components/metrics/SecurePrefsTelemetry.kt:32]: And if want to let this live in a class of its own, mark the invocation function as `suspend` so that dispatcher is set at the callsite, not here.

File: app/src/main/java/org/mozilla/fenix/components/metrics/SecurePrefsTelemetry.kt [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+import mozilla.components.service.nimbus.NimbusApi
+import org.mozilla.fenix.experiments.ExperimentBranch
+import org.mozilla.fenix.experiments.Experiments
+import org.mozilla.fenix.ext.withExperiment
+
+/**
+ * Allows starting a quick test of ACs SecureAbove22Preferences that will emit Facts
+ * for the basic operations and allow us to log them for later evaluation of APIs stability.
+ */
+class SecurePrefsTelemetry(
```
[the remark this answers is not in the document]
Comment (Mugurell) [reply app/src/main/java/org/mozilla/fenix/components/metrics/SecurePrefsTelemetry.kt:21]: If it's okay with you I would prefer keeping `SecurePrefsTelemetry(..).startTests()` just because of the better readability and having a cleaner separation from the rest of the code.

File: app/src/main/java/org/mozilla/fenix/FenixApplication.kt [hunk 8]
```diff
@@ -167,6 +168,7 @@ open class FenixApplication : LocaleAwareApplication(), Provider {
         initVisualCompletenessQueueAndQueueTasks()
 
         components.appStartupTelemetry.onFenixApplicationOnCreate()
+        SecurePrefsTelemetry(this, components.analytics.experiments).startTests()
```
[the remark this answers is not in the document]
Comment (Mugurell) [reply app/src/main/java/org/mozilla/fenix/FenixApplication.kt:171]: Sure. Thank you!

**Comment by Mugurell** [comment]
Thank you Grisha!
I think I've addressed all required changes.

**Comment by Mugurell** [comment]
# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

1) What questions will you answer with this data?
This will help us better assess the stability of our code - the SharedPreferences implementation secured with the Android Keystore.

2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements? Some example responses:
Determine whether the current implementation is stable enough or to better support our users we should look for better solutions.

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?
Crash reports and user reports could serve as an indication of a problem but there is no real alternative.

4) Can current instrumentation answer these questions?
Here we are collecting for the first time specific data about specific APIs.

5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Firefox/Data_Collection) found on the Mozilla wiki.   

Measurement Description | Data Collection Category | Tracking Bug #
-- | -- | --
Exception while running the experiment | category 1 - technical data | #17869 |
Reading a locally stored value fails | category 1 - technical data | #17869 |
Reading a locally stored value succeeds | category 1 - technical data | #17869 |
Storing a value locally fails | category 1 - technical data | #17869 |
Storing a value locally succeeds | category 1 - technical data | #17869 |
Experiment failed and was reset | category 1 - technical data | #17869 |

6) Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
[https://github.com/mozilla-mobile/fenix/blob/master/docs/metrics.md](https://github.com/mozilla-mobile/fenix/blob/master/docs/metrics.md)

7) How long will this data be collected?  Choose one of the following:
This is scoped to a time-limited experiment/project until date 09-01-2021.

8) What populations will you measure?
Ideally all devices with Android API 23+ of all release channels. 
The audience is controlled by a Nimbus experiment which will probably limit the reach.

9) If this data collection is default on, what is the opt-out mechanism for users?
Default Glean SDK opt-out mechanism.

10) Please provide a general description of how you will analyze this data.
Glean / Amplitude

11) Where do you intend to share the results of your analysis?
Only on Glean, Amplitude and with mobile teams.

12) Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection? If so:
 No third-party tool will use this data.

**Comment by Mugurell** [comment]
> Could not determine the dependencies of task ':app:testDebugUnitTest'.
[task 2021-03-05T11:52:21.920Z] > Could not resolve all task dependencies for configuration ':app:debugUnitTestRuntimeClasspath'.
[task 2021-03-05T11:52:21.920Z]    > Could not resolve org.mozilla.components:support-sync-telemetry:74.0.20210303143143.

Restarting CI tasks.

File: app/metrics.yaml [hunk 9]
```diff
@@ -4697,3 +4697,99 @@ awesomebar:
     notification_emails:
       - fenix-core@mozilla.com
     expires: "2021-08-01"
+
+android_keystore_experiment:
+  experiment_failure:
+    type: event
```
Comment (grigoryk) [inline app/metrics.yaml:4703]: `Actions.EXPERIMENT, Values.FAIL` fact comes with an exception too, let's record that in an `extra` like you do for other failures below.

File: app/src/main/java/org/mozilla/fenix/components/metrics/Event.kt [hunk 10]
```diff
@@ -233,6 +234,22 @@ sealed class Event {
             get() = hashMapOf(TopSites.swipeCarouselKeys.page to page.toString())
     }
 
+    object SecurePrefsExperiment : Event()
```
Comment (grigoryk) [inline app/src/main/java/org/mozilla/fenix/components/metrics/Event.kt:237]: `val failureException: String` here too

**Review by grigoryk — COMMENTED** [review]
Thanks! I think this looks correct, just missing the extra keys for the experiment failure fact.

File: app/metrics.yaml (hunk 9 above)
Comment (Mugurell) [reply app/metrics.yaml:4703]: Thank you!
Updated the patch, tested locally this event with one IAE thrown in AC and I got
> {
          "category": "android_keystore_experiment",
          "extra": {
            "failure_exception": "java.lang.IllegalArgumentException"
          },
          "name": "experiment_failure",
          "timestamp": 15
        },

File: app/src/main/java/org/mozilla/fenix/components/metrics/Event.kt (hunk 10 above)
Comment (Mugurell) [reply app/src/main/java/org/mozilla/fenix/components/metrics/Event.kt]: Thank you!
```
     object SecurePrefsExperiment : Event() 
```

now became 
```    
    data class SecurePrefsExperimentFailure(val failureException: String) : Event() {
        override val extras =
            mapOf(AndroidKeystoreExperiment.experimentFailureKeys.failureException to failureException)
    }

**Review by boek — APPROVED** [review]
data review only r+

# Data Review Form (to be filled by Data Stewards)

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way?
Yes, metrics.yaml and metrics.md

2) Is there a control mechanism that allows the user to turn the data collection on and off?
Yes, data collection settings

3) If the request is for permanent data collection, is there someone who will monitor the data over time?
Has expiry

4) Using the **[category system of data types](https://wiki.mozilla.org/Firefox/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?
Type 1

5) Is the data collection request for default-on or default-off?
Default on

6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?
No

7) Is the data collection covered by the existing Firefox privacy notice? **If unsure: escalate to legal if:**
Yes

8) Does there need to be a check-in in the future to determine whether to renew the data?
Fenix team will check-in at expiry

9) Does the data collection use a third-party collection tool?
No

**Comment by Mugurell** [comment]
CI is failing because of
> Could not GET 'https://jcenter.bintray.com/org/gradle/kotlin/plugins/1.3.6/plugins-1.3.6.module'. Received status code 502 from server: Bad Gateway

**Review by grigoryk — APPROVED** [review]

**Comment by grigoryk** [comment]
@Mergifyio backport Mugurell:androidKeystoreExperiment releases_v87.0.0

**Comment by tdsmith** [comment]
Just a flag that when you write result codes into telemetry it would be helpful to include context in the metric descriptions about what those result flags mean or how to find out. Thanks!
